### ISSUE TITLE
Multi Model Connect billing

### ISSUE THREAD

**Issue by syn6UK** [ExpDev07/laravel-cashier-stripe-connect#21]
HI There,

Really love what you've done with the package, one thing I've noticed from your migrations is you only appear to support adding Billable trait to the User model. as your migration only adds the relevant account ID to the users table.

I'm developing a mutli tenant saas platform that allows users to set up multiple stripe connect accounts against each of their models. So someone can have multiple "brands" or storefronts and have stripe connect against each of them.

One suggestion I'd have is building out functionality to allow any model to be Connect billable. I am going to fork this repo and develop it for my own project, but if you want review the changes I've made and consider merging them into a new version, feel free.  I'll also make the repo compatible with Stancl's Laravel Tenancy Package so that other Multi tenant SaaS paltforms can benefit from this.

I'll let you know once i've progressed with this.

**Comment by syn6UK** [comment]
https://github.com/syn6UK/laravel-cashier-stripe-connect

My repo is here.

Given that my changes are quite drastic and it would be a breaking update for most people. I would actually in retrospect discourage a PR here, rather offer my package as a viable alternative for those alternative more complex use cases.

If people are developing a new platform, or using a multi tenant framework like Stancl's Tenancy, my package would be recommended. It has the following:

- UUID Support on Billable Model Primary Key
- Custom Primary Key Support
- Multi Model Supprot (Any model can have Billable trait, not just User)
- Built in support for Tenancy (including publishing tenant migrations

**Comment by ExpDev07** [comment]
Hi! This looks great. Would definitively be nice to have this functionality here, but would have to rethink how to enable multi tenacy support (perhaps through a command or something: php artisan cashier-connect:tenacy) and release as a 2.0 maybe. Your version does not really introduce any breaking changes with a quick glance, but yes, drastic.

**Comment by madewithrunway** [comment]
Totally agree, definitely a 2.0, there will need to be some kind of upgrade / migration artisan command that does the following tasks on an instance:

- Retrieve all users with a stripe account id that isn't null
- Create a mapping record for them using the new schema
- Runs a migration to remove the field from the users table

The upgrade doesn't specifically cater to multi tenancy, rather multi model, so I'd opt for the artisan command syntax of:

php artisan cashier-connect:upgrade which would just run the aforementioned tasks.

I think once that command is built, you could launch this as a V2 with seamless migration. In terms of multi tenancy, nobody currently would be using this package with multi tenancy without already forking it and hammering it into shape, either that or they ripped code out and built their own package, so I think those people may stick with that they have given the work they have done, but offering a 2.0 would be great moving forward.

I'm happy to write that upgrade command on the weekend and this will be ready for an actual release.

In the meantime I'll use my forked package on current projects and swap out for an official release when ready.

**Comment by syn6UK** [comment]
So i've been working on this for the past day or two, I've made a LOT of changes to it, so much so that I think it's better being released as a separate package, as a lot of these will be breaking changes.

I've added

- KYC Webhook support to reduce the number of API calls needed to check for new requirements
- New Routes to support the webhook above
- Direct Charges Feature
- Destination Charges Feature
- Customers Module (WIP) - Allowing the connetced account to manage their customers
- Connetced Products - Allowing connected accounts to manage their products (Ideal for subscriptions)
- Subscriptions -Allowing connected accounts to earn subscription revenue.

I'll be sure to credit everyone involved thus far as this new advnaced package wouldn't have been possible without the ciritcal groundwork already laid.

I'll post a link to the new hosted library when I've finished the last bits.
